src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs

```diff
@@ -148,5 +177,235 @@ private void InitializeBuilder()
+                {
+                    string statusCode = ((HttpWebResponse)webException.Response).StatusCode.ToString();
+                    string statusDesc = ((HttpWebResponse)webException.Response).StatusDescription;
+                    errormsg = string.Format("CosmosDBtrigger status {0}: {1}.", statusCode, statusDesc);
+                }
+                else if (webException != null &&
+                    webException.Status == WebExceptionStatus.NameResolutionFailure)
+                {
+                    errormsg = string.Format("CosmosDBtrigger Exception message: {0}.", webException.Message);
+                }
+                else
+                {
+                    errormsg = e.ToString();
+                }
+
+                _logger.LogWarning(errormsg);
+            }
+
+            return new CosmosDBTriggerMetrics
+            {
+                TimeStamp = DateTime.UtcNow,
+                PartitionCount = partitionCount,
+                RemainingWork = remainingWork
+            };
+        }
+
+        ScaleStatus IScaleMonitor.GetScaleStatus(ScaleStatusContext context)
+        {
+            ScaleStatus status = new ScaleStatus
+            {
+                Vote = ScaleVote.None
+            };
+
+            const int NumberOfSamplesToConsider = 5;
+            List<CosmosDBTriggerMetrics> metrics = (context as ScaleStatusContext<CosmosDBTriggerMetrics>)?.Metrics?.ToList();
+
+            // Unable to determine the correct vote with no metrics.
+            if (metrics == null)
+            {
+                return status;
+            }
+
+            // We shouldn't assign more workers than there are partitions (Cosmos DB, Event Hub, Service Bus Queue/Topic)
+            // This check is first, because it is independent of load or number of samples.
+            int partitionCount = metrics.Last().PartitionCount;
+            if (partitionCount > 0 && partitionCount < context.WorkerCount)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                _logger.LogInformation(string.Format($"WorkerCount ({context.WorkerCount}) > PartitionCount ({partitionCount})."));
+                _logger.LogInformation(string.Format($"Number of instances ({context.WorkerCount}) is too high relative to number " +
+                                                     $"of partitions for collection ({this._monitorCollection.CollectionName}, {partitionCount})."));
+                return status;
+            }
+            
+            // At least 5 samples are required to make a scale decision for the rest of the checks.
+            if (metrics.Count < NumberOfSamplesToConsider)
+            {
+                return status;
+            }
+
+            // Maintain a minimum ratio of 1 worker per 1,000 items of remaining work.
+            long latestRemainingWork = metrics.Last().RemainingWork;
+            if (latestRemainingWork > context.WorkerCount * 1000)
```

Comment:
Would 1000 be a good number? The Trigger can, by default (can be raised), process 100 items per execution, does Event Hub work similarly?

Reply:
I believe Graham has just ported the current scale controller heuristics here. We can tune this as needed in the future, but for now I think parity is fine